File: android/src/main/java/com/brentvatne/exoplayer/RNVPlayerStatisticsListener.java
package com.brentvatne.exoplayer;

import android.os.Handler;
import android.os.Looper;
import android.text.TextUtils;

import androidx.annotation.Nullable;
import androidx.media3.common.ColorInfo;
import androidx.media3.common.Format;
import androidx.media3.common.MediaItem;
import androidx.media3.exoplayer.DecoderReuseEvaluation;
import androidx.media3.exoplayer.analytics.AnalyticsListener;

import com.brentvatne.common.react.VideoEventEmitter;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;

import java.util.Locale;

/**
 * Aggregates playback statistics from Media3 Analytics callbacks and emits a merged, debounced event.
 */
public final class RNVPlayerStatisticsListener implements AnalyticsListener {

    private static final long DEFAULT_DEBOUNCE_MS = 350;

    private final VideoEventEmitter eventEmitter;
    private final Handler handler;

    private boolean enabled;
    private long debounceMs = DEFAULT_DEBOUNCE_MS;
    private boolean posting;

    private final StatsState state = new StatsState();
    
    @Nullable
    private MediaInfo extractedMediaInfo;

    private @Nullable String lastSignature;
    private final Runnable emitRunnable = this::emitIfChanged;

    public RNVPlayerStatisticsListener(VideoEventEmitter eventEmitter) {
        this.eventEmitter = eventEmitter;
        this.handler = new Handler(Looper.getMainLooper());
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
        if (!enabled) {
            handler.removeCallbacks(emitRunnable);
        }
    }

    public void setDebounceMs(long debounceMs) {
        this.debounceMs = Math.max(0, debounceMs);
    }

    public void reset() {
        emitIfChanged();
        handler.removeCallbacks(emitRunnable);
        state.reset();
        lastSignature = null;
    }

    public void setStreamType(@Nullable String streamType) {
        state.streamType = streamType;
        scheduleEmit();
    }
    
    /**
     * Sets the extracted media info as a fallback for missing format information.
     */
    public void setExtractedMediaInfo(@Nullable MediaInfo mediaInfo) {
        this.extractedMediaInfo = mediaInfo;
        scheduleEmit();
    }

    private void scheduleEmit() {
        if (!enabled || posting) {
            return;
        }
        handler.removeCallbacks(emitRunnable);
        handler.postDelayed(emitRunnable, debounceMs);
        this.posting = true;
    }

    private void emitIfChanged() {
        if (!enabled) {
            return;
        }

        this.posting = false;
        WritableMap stats = state.toWritableMap(extractedMediaInfo);
        String signature = state.signature();
        if (signature.equals(lastSignature)) {
            return;
        }
        lastSignature = signature;
        eventEmitter.onVideoStatistics.invoke(stats);
    }

    // --- AnalyticsListener ---

    @Override
    public void onMediaItemTransition(EventTime eventTime, @Nullable MediaItem mediaItem, int reason) {
        // New item => stats should not leak across.
        reset();
    }

    @Override
    public void onAudioDecoderInitialized(EventTime eventTime, String decoderName, long initializedTimestampMs, long initializationDurationMs) {
        state.audioDecoder = decoderName;
        scheduleEmit();
    }

    @Override
    public void onVideoDecoderInitialized(EventTime eventTime, String decoderName, long initializedTimestampMs, long initializationDurationMs) {
        state.videoDecoder = decoderName;
        scheduleEmit();
    }

    @Override
    public void onAudioInputFormatChanged(EventTime eventTime, Format audioFormat, @Nullable DecoderReuseEvaluation decoderReuseEvaluation) {
        state.audioFormat = audioFormat;

        // Container is often present on either format.
        if (!TextUtils.isEmpty(audioFormat.containerMimeType)) {
            state.containerMimeType = audioFormat.containerMimeType;
        }

        scheduleEmit();
    }

    @Override
    public void onVideoInputFormatChanged(EventTime eventTime, Format videoFormat, @Nullable DecoderReuseEvaluation decoderReuseEvaluation) {
        state.videoFormat = videoFormat;

        if (!TextUtils.isEmpty(videoFormat.containerMimeType)) {
            state.containerMimeType = videoFormat.containerMimeType;
        }

        scheduleEmit();
    }

    // --- State + formatting ---

    private static final class StatsState {
        @Nullable String streamType;

        @Nullable String containerMimeType;

        @Nullable Format videoFormat;
        @Nullable String videoDecoder;

        @Nullable Format audioFormat;
        @Nullable String audioDecoder;

        void reset() {
            containerMimeType = null;
            videoFormat = null;
            videoDecoder = null;
            audioFormat = null;
            audioDecoder = null;
        }

        String signature() {
            // Stable, cheap signature to implement "only on change".
            return join(
                    streamType,
                    containerMimeType,
                    videoDecoder,
                    audioDecoder,
                    formatSignature(videoFormat),
                    formatSignature(audioFormat)
            );
        }

        private static @Nullable String formatSignature(@Nullable Format format) {
            if (format == null) return null;
            ColorInfo colorInfo = format.colorInfo;
            return join(
                    format.sampleMimeType,
                    format.codecs,
                    String.valueOf(format.width),
                    String.valueOf(format.height),
                    String.valueOf(format.bitrate),
                    String.valueOf(format.frameRate),
                    String.valueOf(format.channelCount),
                    String.valueOf(format.sampleRate),
                    format.containerMimeType,
                    colorInfo == null ? null : String.valueOf(colorInfo.colorTransfer),
                    colorInfo == null ? null : String.valueOf(colorInfo.colorSpace),
                    colorInfo == null ? null : String.valueOf(colorInfo.colorRange)
            );
        }

        WritableMap toWritableMap() {
            return toWritableMap(null);
        }
        
        WritableMap toWritableMap(@Nullable MediaInfo extractedInfo) {
            WritableMap out = Arguments.createMap();

            // --- New, human-readable fields ---
            out.putString("streamType", streamType);

            String container = VideoMetadataUtils.getContainerDisplayString(containerMimeType);
            if (container != null) {
                out.putString("container", container);
            }

            if (videoFormat != null) {
                out.putString("videoCodecName", VideoMetadataUtils.getCodecDisplayString(videoFormat.sampleMimeType));
                out.putString("resolution", VideoMetadataUtils.getResolutionDisplayString(videoFormat.width, videoFormat.height));
                
                // Use extracted media info framerate as fallback if ExoPlayer format has no framerate
                float frameRate = videoFormat.frameRate;
                if (frameRate <= 0 && extractedInfo != null && extractedInfo.getFrameRate() > 0) {
                    frameRate = extractedInfo.getFrameRate();
                }
                out.putString("frameRate", VideoMetadataUtils.getFrameRateDisplayString(frameRate));
                
                out.putString("decodedVideoFormat", getDecodedFormatDisplayString(videoFormat));
            } else if (extractedInfo != null) {
                // Use extracted info if no ExoPlayer format available
                out.putString("resolution", VideoMetadataUtils.getResolutionDisplayString(
                    (int) extractedInfo.getVideoWidth(), 
                    (int) extractedInfo.getVideoHeight()
                ));
                out.putString("frameRate", VideoMetadataUtils.getFrameRateDisplayString(extractedInfo.getFrameRate()));
            }

            if (audioFormat != null) {
                out.putString("audioCodecName", VideoMetadataUtils.getCodecDisplayString(audioFormat.sampleMimeType));
                out.putString("decodedAudioFormat", getDecodedFormatDisplayString(audioFormat));
                out.putString("audioLayout", VideoMetadataUtils.getAudioLayoutDisplayString(audioFormat.channelCount));
                out.putString("decodedAudioChannels", VideoMetadataUtils.getAudioChannelsDisplayString(audioFormat.channelCount));
            }

            out.putString("videoDecoder", videoDecoder);
            out.putString("audioDecoder", audioDecoder);

            // Use extracted bitrate as fallback
            int videoBitrate = videoFormat != null ? videoFormat.bitrate : Format.NO_VALUE;
            if (videoBitrate <= 0 && extractedInfo != null && extractedInfo.getBitrate() > 0) {
                videoBitrate = (int) extractedInfo.getBitrate();
            }
            
            out.putString(
                    "bitrate",
                    VideoMetadataUtils.getCombinedBitrateDisplayString(
                            videoBitrate,
                            audioFormat != null ? audioFormat.bitrate : Format.NO_VALUE
                    )
            );

            out.putString("profileLevel", getProfileLevelSummary());

            return out;
        }

        private @Nullable String getProfileLevelSummary() {
            if (videoFormat == null) {
                return null;
            }

            String dynamicRange = VideoMetadataUtils.getDynamicRangeDisplayName(videoFormat.sampleMimeType, videoFormat.colorInfo);

            android.util.Pair<Integer, Integer> profileLevel = null;
            try {
                profileLevel = androidx.media3.exoplayer.mediacodec.MediaCodecUtil.getCodecProfileAndLevel(videoFormat);
            } catch (Exception ignored) {
                // ignore
            }

            String codecProfileLevel = null;
            if (profileLevel != null) {
                codecProfileLevel = VideoMetadataUtils.getCodecProfileLevelDisplayString(
                        videoFormat.sampleMimeType,
                        profileLevel.first,
                        profileLevel.second
                );
            }

            if (!TextUtils.isEmpty(codecProfileLevel) && !TextUtils.isEmpty(dynamicRange)) {
                return String.format(Locale.US, "%s, %s", dynamicRange, codecProfileLevel);
            }
            if (!TextUtils.isEmpty(codecProfileLevel)) {
                return codecProfileLevel;
            }
            if (!TextUtils.isEmpty(dynamicRange)) {
                return dynamicRange;
            }
            return null;
        }

        private static @Nullable String getDecodedFormatDisplayString(Format format) {
            if (format == null) return null;

            String codec = VideoMetadataUtils.getCodecDisplayString(format.sampleMimeType);

            android.util.Pair<Integer, Integer> profileLevel = null;
            try {
                profileLevel = androidx.media3.exoplayer.mediacodec.MediaCodecUtil.getCodecProfileAndLevel(format);
            } catch (Exception ignored) {
                // ignore
            }

            if (profileLevel == null) {
                return codec;
            }

            String pl = VideoMetadataUtils.getCodecProfileLevelDisplayString(
                    format.sampleMimeType,
                    profileLevel.first,
                    profileLevel.second
            );
            if (TextUtils.isEmpty(pl)) {
                return codec;
            }
            return codec + ", " + pl;
        }

        private static String join(@Nullable String... parts) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < parts.length; i++) {
                if (i > 0) sb.append('\u0001');
                sb.append(parts[i] == null ? "" : parts[i]);
            }
            return sb.toString();
        }
    }
}

File: packages/react-native-video/android/src/main/java/com/twg/video/core/player/RNVPlayerStatisticsListener.kt
package com.twg.video.core.player

import android.os.Handler
import android.os.Looper
import android.util.Pair
import androidx.annotation.Nullable
import androidx.media3.common.ColorInfo
import androidx.media3.common.Format
import androidx.media3.common.MediaItem
import androidx.media3.exoplayer.DecoderReuseEvaluation
import androidx.media3.exoplayer.analytics.AnalyticsListener
import androidx.media3.exoplayer.mediacodec.MediaCodecUtil
import com.margelo.nitro.video.HybridVideoPlayerEventEmitter
import com.margelo.nitro.video.VideoStatisticsData

/**
 * Aggregates playback statistics from Media3 Analytics callbacks and emits a merged,
 * debounced `onVideoStatistics` event.
 *
 * <p>Ported from the DodoStream fork (v6).
 */
class RNVPlayerStatisticsListener(
  private val eventEmitter: HybridVideoPlayerEventEmitter,
) : AnalyticsListener {

  private companion object {
    const val TAG = "RNVPlayerStatisticsListener"
    const val DEFAULT_DEBOUNCE_MS = 350L
  }

  private val handler = Handler(Looper.getMainLooper())

  private var enabled = false
  private var debounceMs = DEFAULT_DEBOUNCE_MS
  private var posting = false

  private val state = StatsState()

  @Nullable
  private var extractedMediaInfo: MediaInfo? = null

  private var lastSignature: String? = null

  private val emitRunnable = Runnable { emitIfChanged() }

  fun setEnabled(enabled: Boolean) {
    this.enabled = enabled
    if (!enabled) {
      handler.removeCallbacks(emitRunnable)
    }
  }

  fun setDebounceMs(debounceMs: Long) {
    this.debounceMs = debounceMs.coerceAtLeast(0)
  }

  fun reset() {
    emitIfChanged()
    handler.removeCallbacks(emitRunnable)
    state.reset()
    lastSignature = null
  }

  fun setStreamType(streamType: String?) {
    state.streamType = streamType
    scheduleEmit()
  }

  /** Sets the extracted media info as a fallback for missing format information. */
  fun setExtractedMediaInfo(mediaInfo: MediaInfo?) {
    this.extractedMediaInfo = mediaInfo
    scheduleEmit()
  }

  private fun scheduleEmit() {
    if (!enabled || posting) {
      return
    }
    handler.removeCallbacks(emitRunnable)
    handler.postDelayed(emitRunnable, debounceMs)
    posting = true
  }

  private fun emitIfChanged() {
    if (!enabled) {
      return
    }

    posting = false
    val signature = state.signature()
    if (signature == lastSignature) {
      return
    }
    lastSignature = signature
    eventEmitter.onVideoStatistics(state.toVideoStatisticsData(extractedMediaInfo))
  }

  // --- AnalyticsListener ---

  override fun onMediaItemTransition(
    eventTime: AnalyticsListener.EventTime,
    mediaItem: MediaItem?,
    reason: Int,
  ) {
    // New item => stats should not leak across.
    reset()
  }

  override fun onAudioDecoderInitialized(
    eventTime: AnalyticsListener.EventTime,
    decoderName: String,
    initializedTimestampMs: Long,
    initializationDurationMs: Long,
  ) {
    state.audioDecoder = decoderName
    scheduleEmit()
  }

  override fun onVideoDecoderInitialized(
    eventTime: AnalyticsListener.EventTime,
    decoderName: String,
    initializedTimestampMs: Long,
    initializationDurationMs: Long,
  ) {
    state.videoDecoder = decoderName
    scheduleEmit()
  }

  override fun onAudioInputFormatChanged(
    eventTime: AnalyticsListener.EventTime,
    format: Format,
    decoderReuseEvaluation: DecoderReuseEvaluation?,
  ) {
    state.audioFormat = format

    // Container is often present on either format.
    if (!format.containerMimeType.isNullOrEmpty()) {
      state.containerMimeType = format.containerMimeType
    }

    scheduleEmit()
  }

  override fun onVideoInputFormatChanged(
    eventTime: AnalyticsListener.EventTime,
    format: Format,
    decoderReuseEvaluation: DecoderReuseEvaluation?,
  ) {
    state.videoFormat = format

    if (!format.containerMimeType.isNullOrEmpty()) {
      state.containerMimeType = format.containerMimeType
    }

    scheduleEmit()
  }

  // --- State + formatting ---

  private class StatsState {
    @Nullable
    var streamType: String? = null

    @Nullable
    var containerMimeType: String? = null

    @Nullable
    var videoFormat: Format? = null
    @Nullable
    var videoDecoder: String? = null

    @Nullable
    var audioFormat: Format? = null
    @Nullable
    var audioDecoder: String? = null

    fun reset() {
      containerMimeType = null
      videoFormat = null
      videoDecoder = null
      audioFormat = null
      audioDecoder = null
    }

    fun signature(): String {
      // Stable, cheap signature to implement "only on change".
      return join(
        streamType,
        containerMimeType,
        videoDecoder,
        audioDecoder,
        formatSignature(videoFormat),
        formatSignature(audioFormat),
      )
    }

    private fun formatSignature(format: Format?): String? {
      if (format == null) return null
      val colorInfo = format.colorInfo
      return join(
        format.sampleMimeType,
        format.codecs,
        format.width.toString(),
        format.height.toString(),
        format.bitrate.toString(),
        format.frameRate.toString(),
        format.channelCount.toString(),
        format.sampleRate.toString(),
        format.containerMimeType,
        colorInfo?.colorTransfer?.toString(),
        colorInfo?.colorSpace?.toString(),
        colorInfo?.colorRange?.toString(),
      )
    }

    fun toVideoStatisticsData(extractedInfo: MediaInfo?): VideoStatisticsData {
      val videoFormat = videoFormat
      val audioFormat = audioFormat

      var frameRate: String? = null
      var resolution: String? = null
      var bitrate: String? = null

      if (videoFormat != null) {
        // Use extracted media info framerate as fallback if ExoPlayer format has no framerate
        var fps = videoFormat.frameRate
        if (fps <= 0 && extractedInfo != null && extractedInfo.frameRate > 0) {
          fps = extractedInfo.frameRate
        }
        frameRate = VideoMetadataUtils.getFrameRateDisplayString(fps)
        resolution = VideoMetadataUtils.getResolutionDisplayString(videoFormat.width, videoFormat.height)

        // Use extracted bitrate as fallback
        var videoBitrate = videoFormat.bitrate
        if (videoBitrate <= 0 && extractedInfo != null && extractedInfo.bitrate > 0) {
          videoBitrate = extractedInfo.bitrate.toInt()
        }
        bitrate = VideoMetadataUtils.getCombinedBitrateDisplayString(
          videoBitrate,
          audioFormat?.bitrate ?: Format.NO_VALUE,
        )
      } else if (extractedInfo != null) {
        // Use extracted info if no ExoPlayer format available
        resolution = VideoMetadataUtils.getResolutionDisplayString(
          extractedInfo.videoWidth.toInt(),
          extractedInfo.videoHeight.toInt(),
        )
        frameRate = VideoMetadataUtils.getFrameRateDisplayString(extractedInfo.frameRate)
      }

      return VideoStatisticsData(
        streamType = streamType,
        container = VideoMetadataUtils.getContainerDisplayString(containerMimeType),
        videoCodecName = videoFormat?.let { VideoMetadataUtils.getCodecDisplayString(it.sampleMimeType) },
        audioCodecName = audioFormat?.let { VideoMetadataUtils.getCodecDisplayString(it.sampleMimeType) },
        resolution = resolution,
        frameRate = frameRate,
        bitrate = bitrate,
        profileLevel = getProfileLevelSummary(videoFormat),
        decodedVideoFormat = videoFormat?.let { getDecodedFormatDisplayString(it) },
        decodedAudioFormat = audioFormat?.let { getDecodedFormatDisplayString(it) },
        decodedAudioChannels = audioFormat?.let { VideoMetadataUtils.getAudioChannelsDisplayString(it.channelCount) },
        audioLayout = audioFormat?.let { VideoMetadataUtils.getAudioLayoutDisplayString(it.channelCount) },
        videoDecoder = videoDecoder,
        audioDecoder = audioDecoder,
      )
    }

    private fun getProfileLevelSummary(format: Format?): String? {
      if (format == null) {
        return null
      }

      val dynamicRange = VideoMetadataUtils.getDynamicRangeDisplayName(format.sampleMimeType, format.colorInfo)

      val profileLevel: Pair<Int, Int>? = try {
        MediaCodecUtil.getCodecProfileAndLevel(format)
      } catch (ignored: Exception) {
        null
      }

      val codecProfileLevel: String? = if (profileLevel != null) {
        VideoMetadataUtils.getCodecProfileLevelDisplayString(
          format.sampleMimeType,
          profileLevel.first,
          profileLevel.second,
        )
      } else {
        null
      }

      if (!codecProfileLevel.isNullOrEmpty() && !dynamicRange.isNullOrEmpty()) {
        return "$dynamicRange, $codecProfileLevel"
      }
      if (!codecProfileLevel.isNullOrEmpty()) {
        return codecProfileLevel
      }
      return dynamicRange
    }

    private fun getDecodedFormatDisplayString(format: Format): String? {
      val codec = VideoMetadataUtils.getCodecDisplayString(format.sampleMimeType)

      val profileLevel: Pair<Int, Int>? = try {
        MediaCodecUtil.getCodecProfileAndLevel(format)
      } catch (ignored: Exception) {
        null
      }

      if (profileLevel == null) {
        return codec
      }

      val pl = VideoMetadataUtils.getCodecProfileLevelDisplayString(
        format.sampleMimeType,
        profileLevel.first,
        profileLevel.second,
      )
      if (pl.isNullOrEmpty()) {
        return codec
      }
      return "$codec, $pl"
    }

    private fun join(vararg parts: String?): String {
      val sb = StringBuilder()
      for ((i, part) in parts.withIndex()) {
        if (i > 0) sb.append('\u0001')
        sb.append(part ?: "")
      }
      return sb.toString()
    }
  }
}
